source/openqasm/parser/antlr/qasm_parser.py

```diff
@@ -0,0 +1,893 @@
+            add_span(
+                Qubit(
+                    ctx.Identifier().getText(),
+                ),
+                get_span(ctx.Identifier()),
+            ),
+            self.visit(ctx.designator()) if ctx.designator() else None,
+        )
+
+    @span
+    def visitDesignator(self, ctx: qasm3Parser.DesignatorContext):
+        return self.visit(ctx.expression())
+
+    @span
+    def visitStatement(self, ctx: qasm3Parser.StatementContext):
+        return self.visit(ctx.getChild(0))
+
+    @span
+    def visitAliasStatement(self, ctx: qasm3Parser.AliasStatementContext):
+        target = Identifier(name=ctx.Identifier().getText())
+        value = self.visit(ctx.indexIdentifier())
+        return AliasStatement(add_span(target, get_span(ctx.Identifier())), value)
+
+    @span
+    def visitExpressionStatement(self, ctx: qasm3Parser.ExpressionStatementContext):
+        return ExpressionStatement(self.visit(ctx.expression()))
+
+    @span
+    def visitQuantumStatement(self, ctx: qasm3Parser.QuantumStatementContext):
+        return self.visit(ctx.getChild(0))
+
+    @span
+    def visitQuantumInstruction(self, ctx: qasm3Parser.QuantumInstructionContext):
+        quantum_gate_call = ctx.quantumGateCall()
+        quantum_phase = ctx.quantumPhase()
+        quantum_measurement = ctx.quantumMeasurement()
+        quantum_reset = ctx.quantumReset()
+        quantum_barrier = ctx.quantumBarrier()
+        if quantum_gate_call:
+            return self.visit(quantum_gate_call)
+        elif quantum_phase:
+            return self.visit(quantum_phase)
+        elif quantum_measurement:
+            return self.visit(quantum_measurement)
+        elif quantum_reset:
+            return self.visit(quantum_reset)
+        elif quantum_barrier:
+            return self.visit(quantum_barrier)
+        else:
+            raise NotImplementedError(
+                f"Not implemented QuantumInstructionContext at {get_span(ctx)}"
+            )
+
+    @span
+    def visitQuantumGateCall(self, ctx: qasm3Parser.QuantumGateCallContext):
+        modifiers = [self.visit(modifier) for modifier in ctx.quantumGateModifier()]
+        gate_name = ctx.quantumGateName().getText()
+        expression_list = ctx.expressionList()
+        if expression_list:
+            arguments = [self.visit(expression) for expression in expression_list.expression()]
+        else:
+            arguments = []
+        qubits = [self.visit(qubit) for qubit in ctx.indexIdentifierList().indexIdentifier()]
+
+        gate = QuantumGate(modifiers=modifiers, name=gate_name, arguments=arguments, qubits=qubits)
+
+        return gate
+
+    @span
+    def visitQuantumPhase(self, ctx: qasm3Parser.QuantumPhaseContext):
+        modifiers = [self.visit(modifier) for modifier in ctx.quantumGateModifier()]
+        argument = self.visit(ctx.expression())
+        qubits = [self.visit(qubit) for qubit in ctx.indexIdentifierList().indexIdentifier()]
```

Comment:
This line fails on the following OpenQASM line:
```
gate rz(λ) a { gphase(2); U(0, 0, λ) a; }
```
because `ctx.indexIdentifierList()` is `None` on the `gphase(2);` part.
This OpenQASM 3 line is valid according to [https://qiskit.github.io/openqasm/language/gates.html](https://qiskit.github.io/openqasm/language/gates.html?highlight=gphase) and the ANTLR reference grammar. You can also find a more complex version of it in `stdgates.inc`.

Reply:
Updated.